lib/winapi/sync.c

```diff
@@ -1,11 +1,145 @@
+        return FALSE;
+    }
+
+    int oldCount = atomic_fetch_sub(&ConditionVariable->waitCount, 1);
+    if (oldCount == 1) {
+        // Last waiter needs to manually reset the broadcast-object
+        status = NtClearEvent(ConditionVariable->eventHandles[1]);
+        if (!NT_SUCCESS(status)) {
+            EnterCriticalSection(CriticalSection);
+            return FALSE;
+        }
+    }
+
+    EnterCriticalSection(CriticalSection);
+    return TRUE;
+}
+
+BOOL SleepConditionVariableSRW (PCONDITION_VARIABLE ConditionVariable, PSRWLOCK SRWLock, DWORD dwMilliseconds, ULONG Flags)
+{
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
+
+    NTSTATUS status;
+    LARGE_INTEGER waitTime;
+    LARGE_INTEGER *waitTimePtr = NULL;
+
+    if (dwMilliseconds != INFINITE) {
+        waitTime.QuadPart = dwMilliseconds * 1000;
+        waitTimePtr = &waitTime;
+    }
+
+    if (Flags == CONDITION_VARIABLE_LOCKMODE_SHARED) {
+        ReleaseSRWLockShared(SRWLock);
+    } else {
+        ReleaseSRWLockExclusive(SRWLock);
+    }
+    atomic_fetch_add(&ConditionVariable->waitCount, 1);
+    status = NtWaitForMultipleObjectsEx(2, ConditionVariable->eventHandles, WaitAny, UserMode, FALSE, waitTimePtr);
+    if (status == WAIT_FAILED || status == STATUS_TIMEOUT) {
+        atomic_fetch_sub(&ConditionVariable->waitCount, 1);
+        if (Flags == CONDITION_VARIABLE_LOCKMODE_SHARED) {
+            AcquireSRWLockShared(SRWLock);
+        } else {
+            AcquireSRWLockExclusive(SRWLock);
+        }
+        return FALSE;
+    }
+
+    int oldCount = atomic_fetch_sub(&ConditionVariable->waitCount, 1);
+    if (oldCount == 1) {
+        // Last waiter needs to manually reset the broadcast-object
+        status = NtClearEvent(ConditionVariable->eventHandles[1]);
+        if (!NT_SUCCESS(status)) {
+            if (Flags == CONDITION_VARIABLE_LOCKMODE_SHARED) {
+                AcquireSRWLockShared(SRWLock);
+            } else {
+                AcquireSRWLockExclusive(SRWLock);
+            }
+            return FALSE;
+        }
+    }
+
+    if (Flags == CONDITION_VARIABLE_LOCKMODE_SHARED) {
+        AcquireSRWLockShared(SRWLock);
+    } else {
+        AcquireSRWLockExclusive(SRWLock);
+    }
+    return TRUE;
+}
+
+VOID WakeConditionVariable (PCONDITION_VARIABLE ConditionVariable)
+{
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
+
+    NTSTATUS status;
+    status = NtSetEvent(ConditionVariable->eventHandles[0], NULL);
+    assert(NT_SUCCESS(status));
+}
+
+VOID WakeAllConditionVariable (PCONDITION_VARIABLE ConditionVariable)
+{
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
+
+    NTSTATUS status;
+    status = NtSetEvent(ConditionVariable->eventHandles[1], NULL);
+    assert(NT_SUCCESS(status));
+}
+
+VOID DestroyConditionVariable (PCONDITION_VARIABLE ConditionVariable)
```

Comment:
`UninitializeConditionVariable` would sound more logical to me. Initialization creates the events; so `Uninitialize` would be a counterpart.
`Destroy` would, to me, imply that there'd be a matching `Create` somewhere, which allocated that memory (which is not the case).